### PR TITLE
HSEARCH-3490 Add experimental support for Elasticsearch 7.0.0.beta1

### PR DESCRIPTION
**Pull request by yrodiere**
https://hibernate.atlassian.net//browse/HSEARCH-3490

I ran a full build (ES5, ES6, ES7) there: http://ci.hibernate.org/blue/organizations/jenkins/hibernate-search-yoann/detail/HSEARCH-3490/4/

### CODE REVIEW

**Comment by yrodiere** [comment]
Force-pushed an update to address some Sonar warnings.

Note the loss in coverage is a bit artificial: coverage numbers only takes into account the "default" build, i.e. the one with ES6, not the one with ES7.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/dialect/impl/es6/Elasticsearch6Dialect.java [hunk 1]
```diff
@@ -7,28 +7,38 @@
...
 
 	@Override
-	public Elasticsearch6WorkBuilderFactory createWorkBuilderFactory(GsonProvider gsonProvider) {
-		/*
-		 * The ES6 factory works fine with ES5 so far.
-		 * We may have to override some methods in the future if we add new work types, though.
-		 */
-		return new Elasticsearch6WorkBuilderFactory( gsonProvider );
+	public ElasticsearchSearchResultExtractorFactory createSearchResultExtractorFactory() {
+		// Necessary because the total hit count it formatted differently in ES7
```
Comment (gsmet) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/dialect/impl/es6/Elasticsearch6Dialect.java]: s/it formatted/is formatted/

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/dialect/impl/es7/Elasticsearch7Dialect.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import org.hibernate.search.backend.elasticsearch.types.dsl.provider.impl.Elasticsearch7IndexFieldTypeFactoryContextProvider;
+import org.hibernate.search.backend.elasticsearch.types.dsl.provider.impl.ElasticsearchIndexFieldTypeFactoryContextProvider;
+import org.hibernate.search.backend.elasticsearch.work.builder.factory.impl.Elasticsearch7WorkBuilderFactory;
+import org.hibernate.search.backend.elasticsearch.work.builder.factory.impl.ElasticsearchWorkBuilderFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * The dialect for Elasticsearch 6.
```
Comment (gsmet) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/dialect/impl/es7/Elasticsearch7Dialect.java]: 7

**Comment by yrodiere** [comment]
I addressed your comments, rebased and force-pushed. Waiting for CI because of the rebase.

> A bit worried by the amount of work required to support a new version of ES. But I suppose that's how it is, considering how they deal with backward compatibility.

On the bright side, the actual changes are only `813 insertions(+), 456 deletions(-)`, the rest being tests and test resources (init scripts, ...), and even for tests there's the introduction of test dialects which will hopefully represent less work next time, now that it's there.

**Comment by yrodiere** [comment]
Merged, thanks!

**Pull request merged**